lib/tpm/tpm.go

```diff
@@ -0,0 +1,216 @@
+	ctx, span := tracer.Start(ctx, "Query")
+	defer span.End()
+
+	tpm, err := attest.OpenTPM(&attest.OpenConfig{
+		TPMVersion: attest.TPMVersion20,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		if err := tpm.Close(); err != nil {
+			log.ErrorContext(
+				ctx,
+				"Failed to close TPM",
+				slog.String("error", err.Error()),
+			)
+		}
+	}()
+	return queryWithTPM(ctx, log, tpm)
+}
+
+func queryWithTPM(
+	ctx context.Context, log *slog.Logger, tpm *attest.TPM,
+) (*QueryRes, error) {
+	ctx, span := tracer.Start(ctx, "queryWithTPM")
+	defer span.End()
+
+	data := &QueryRes{}
+
+	eks, err := tpm.EKs()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(eks) == 0 {
+		// This is a pretty unusual case, `go-attestation` will attempt to
+		// create an EK if no EK Certs are present in the NVRAM of the TPM.
+		// Either way, it lets us catch this early in case `go-attestation`
+		// misbehaves.
+		return nil, trace.BadParameter(
+			"no endorsement keys found in tpm",
+		)
+	}
+
+	// The first EK returned by `go-attestation` will be an RSA based EK key or
+	// EK cert. On Windows, ECC certs may also be returned following this. At
+	// this time, we are only interested in RSA certs, so we just consider the
+	// first thing returned.
+	ekPub, err := x509.MarshalPKIXPublicKey(eks[0].Public)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	data.EKPub = ekPub
+	data.EKPubHash, err = hashEKPub(eks[0].Public)
+	if err != nil {
+		return nil, trace.Wrap(err, "hashing ekpub")
+	}
+
+	if eks[0].Certificate != nil {
+		data.EKCert = eks[0].Certificate.Raw
+		data.EKCertSerial = serialString(eks[0].Certificate.SerialNumber)
+		data.EKCertPresent = true
+	}
+	log.DebugContext(ctx, "Successfully queried TPM", "data", data)
+	return data, nil
+}
+
+// Attest provides the information necessary to perform a tpm join to a Teleport
+// cluster. It returns a solve function which should be called when the
+// encrypted credential challenge is received from the server.
+// The Close function must be called if Attest returns in a non-error state.
+func Attest(ctx context.Context, log *slog.Logger) (
+	data *QueryRes,
+	attestParams *attest.AttestationParameters,
+	solve func(ec *attest.EncryptedCredential) ([]byte, error),
+	close func() error,
+	err error,
+) {
+	ctx, span := tracer.Start(ctx, "Attest")
+	defer span.End()
+
+	tpm, err := attest.OpenTPM(&attest.OpenConfig{
+		TPMVersion: attest.TPMVersion20,
+	})
+	if err != nil {
+		return nil, nil, nil, nil, trace.Wrap(err)
+	}
+	return attestWithTPM(ctx, log, tpm)
+
+}
+
+func attestWithTPM(ctx context.Context, log *slog.Logger, tpm *attest.TPM) (
+	data *QueryRes,
+	attestParams *attest.AttestationParameters,
+	solve func(ec *attest.EncryptedCredential) ([]byte, error),
+	close func() error,
+	err error,
+) {
+	ctx, span := tracer.Start(ctx, "attestWithTPM")
+	defer span.End()
+	defer func() {
+		if err != nil {
+			if err := tpm.Close(); err != nil {
```

Comment:
Yeah - that makes sense. I originally just had the top level functions but for the purpose of testing I needed to be able to inject the TPM. Definitely makes more sense for the closure logic to be in Attest rather than attestWithTPm.